src/lua/uri.c

```diff
@@ -0,0 +1,313 @@
+		lua_rawgeti(L, -1, i + 1);
+		nested_tables_is_present = (lua_type(L, -1) == LUA_TTABLE);
+		lua_pop(L, 1);
+	}
+	return nested_tables_is_present;
+}
+
+/**
+ * Perform some checks for the URI table. The types of checks depend on
+ * @a table_is_nested.
+ */
+static bool
+uri_table_is_valid(struct lua_State *L, bool table_is_nested)
+{
+	/** All numerical keys in URI table corresponds to URI */
+	int uri_count_from_numerical_keys = lua_objlen(L, -1);
+	bool params_is_present = is_field_present(L, "params");
+	bool uri_key_is_present = is_field_present(L, "uri");
+	bool nested_uri_tables_is_present = is_nested_uri_tables_present(L);
+	bool default_params_is_present = is_field_present(L, "default_params");
+
+        /* Should be at least one URI in URI table */
+	if (uri_count_from_numerical_keys == 0 && !uri_key_is_present)
+		return false;
+	/* Mix "uri=" syntax and URIs from numerical keys is banned */
+	if (uri_count_from_numerical_keys != 0 && uri_key_is_present)
+		return false;
+	/* Should be no more than one URI in nested URI table */
+	if (uri_count_from_numerical_keys > 1 && table_is_nested)
+		return false;
+	/* Table "params" is now allowed for several URIs */
+	if (uri_count_from_numerical_keys > 1 && params_is_present)
+		return false;
+	/* Nested tables is not allowed in nested tables */
+	if (table_is_nested && nested_uri_tables_is_present)
+		return false;
+	/* Table "default_params" is not allowed in nested tables */
+	if (default_params_is_present && table_is_nested)
+		return false;
+	/*
+	 * Table "params" is not allowed in table, which contains,
+	 * nested URI tables.
+	 */
+	if (params_is_present && nested_uri_tables_is_present)
+		return false;
+	return true;
+}
+
+/**
+ * Create @a uri_set from the table, which located at the top of lua stack.
+ * Supports many different formats.
+ */
+static int
+uri_set_create_from_lua(struct uri_set *uri_set, struct lua_State *L,
+			bool table_is_nested)
+{
+	int rc = 0;
+	struct uri_set u = { 0 };
+
+	/** Some preliminary checks of the URI table. */
+	if (!uri_table_is_valid(L, table_is_nested))
+		return -1;
+
+	lua_pushnil(L);
+	while (lua_next(L, lua_gettop(L) - 1) != 0 && rc == 0) {
+		int key_type = lua_type(L, -2);
+		int val_type = lua_type(L, -1);
+		const char *key;
+		const char *val;
+		switch (key_type) {
+		case LUA_TNUMBER:
+			switch (val_type) {
+			case LUA_TSTRING:
+			case LUA_TNUMBER:
+				val = lua_tostring(L, -1);
+				rc = uri_set_add_uri(&u, val);
+				break;
+			case LUA_TTABLE:
+				rc = uri_set_create_from_lua(&u, L, true);
+				break;
+			default:
+				rc = -1;
+			};
+			break;
+		case LUA_TSTRING:
+			key = lua_tostring(L, -2);
+			if (strcmp(key, "uri") == 0) {
+				switch (val_type) {
+				case LUA_TSTRING:
+				case LUA_TNUMBER:
+					val = lua_tostring(L, -1);
+					rc = uri_set_add_uri(&u, val);
+					break;
+				default:
+					rc = -1;
+				}
+			} else if (strcmp(key, "default_params") != 0 &&
+				   strcmp(key, "params") != 0) {
+				rc = -1;
```

Comment:
Let's ignore unknown keys. This is conventional.

Reply:
fixed